src/math/PercentageMath.sol

```diff
@@ -100,45 +99,41 @@ library PercentageMath {
         }
     }
 
-    /// @notice Executes a weighted average, given an interval [x, y] and a percent p: x * (1 - p) + y * p
-    /// @param x The value at the start of the interval (included).
-    /// @param y The value at the end of the interval (included).
-    /// @param percentage The percentage of the interval to be calculated.
-    /// @return z The average of x and y, weighted by percentage.
+    /// @notice Executes a weighted average (x * (1 - p) + y * p), rounded up.
+    /// @param x The first value, with a weight of 1 - percentage.
+    /// @param y The second value, with a weight of percentage.
+    /// @param percentage The weight of y, and complement of the weight of x.
+    /// @return z The result of the weighted average.
     function weightedAvg(
         uint256 x,
         uint256 y,
         uint256 percentage
     ) internal pure returns (uint256 z) {
-        // Let percentage > 0
-        // Underflow if percentage > PERCENTAGE_FACTOR
-        //
-        // Overflow if x1 = x * (PERCENTAGE_FACTOR - percentage) + HALF_PERCENTAGE_FACTOR > type(uint256).max
-        // <=> x * (PERCENTAGE_FACTOR - percentage) > type(uint256).max - HALF_PERCENTAGE_FACTOR
-        // <=> x > (type(uint256).max - HALF_PERCENTAGE_FACTOR) / (PERCENTAGE_FACTOR - percentage)
-        //
-        // Overflow if x2 = y * percentage + HALF_PERCENTAGE_FACTOR > type(uint256).max
+        // Must revert if
+        // percentage > PERCENTAGE_FACTOR
+        // or
+        // x1 = x * (PERCENTAGE_FACTOR - percentage) + HALF_PERCENTAGE_FACTOR > type(uint256).max
+        // <=> (PERCENTAGE_FACTOR - percentage) > 0 and x > (type(uint256).max - HALF_PERCENTAGE_FACTOR) / (PERCENTAGE_FACTOR - percentage)
+        // or
+        // x2 = y * percentage + HALF_PERCENTAGE_FACTOR > type(uint256).max
         // <=> y * percentage > type(uint256).max - HALF_PERCENTAGE_FACTOR
-        // <=> y > (type(uint256).max - HALF_PERCENTAGE_FACTOR) / percentage
+        // <=> percentage > 0 and y > (type(uint256).max - HALF_PERCENTAGE_FACTOR) / percentage
         //
-        // Overflow if x1 + x2 > type(uint256).max
+        // Note: x1 / PERCENTAGE_FACTOR + x2 / PERCENTAGE_FACTOR < max(x,y) <= type(uint256).max
```

Comment:
Let's write `PF` for `PERCENTAGE_FACTOR` and `HPF` for `HALF_PERCENTAGE_FACTOR`.

I don't think this note is true, for example this quantity is equal to the max when `x = y` and `percentage = PF`.
It can even be strictly greater than the max. For example with the values `x = y = 5000` and `percentage = 1`, we have `x1 / PF == x` and `x2 / PF == 1`

I still think there is no overflow when adding at the end (so this PR is safe), for the following reason. Let's assume that `percentage` is not equal to 0 and not equal to `PF`, otherwise it's obvious that there is no overflow because we would add 0. Then, because we checked overflow on `x1` and `x2`, we know that `x <= type(uint256).max - HPF` and same thing for `y`. On the other hand, we have a upper bound for `x1 / PF` given by `x * (1-p) + 1` where this last value is given in $\mathbb{R}$. This is obtained by changing `HPF` into `PF` in `x1`. Similarly, we have `x2 / PF` that is lower than `y * p +1`.

So `x1 / PF + x2 / PF <= x * (1-p) + 1 + y * p + 1 <= max(x, y) + 2 <= type(uint256).max - HPF + 2 <= type(uint256).max`

Reply:
Ok, great that you dived on this, I didn't have in mind this edge case. 
But it makes me realize that this function might have an unwanted behavior : weighted average of a and b is not in [a,b]